### ISSUE TITLE
RESTful API for remote management

### ISSUE THREAD

**Issue by limikael** [afragen/github-updater#354]
I like the remote management feature, but unfortunately I don't use "iThemes Sync" or "InfiniteWP". I'm sure they are great, but I'm kind of stuck in liking my command line. Therefore, it would be great if it would be possible to update a remote site by accessing a RESTful api on the site. That way I can create a script that would update a remote site using curl or something similar. I saw there is some discussion here about supporting wp-cli, but that doesn't really solve my issue, since wp-cli has to be run on the actual server where the site is run. I would like to be able to run an update script to update a remote server.

Please see this as the opening of a dialog. I would be happy to take a stab at creating a PR for this feature. I would like to ask if such a PR would be well received.

And thanks for a great plugin!

**Comment by afragen** [comment]
@limikael I'm all in favor of a PR for REST API endpoints but how would you authenticate to allow for the updates?

Before starting a PR I would really like to know how it would work.

**Comment by limikael** [comment]
Good question!

I envision it like this. On the "Remote Management" tab, there would be a checkbox to enable the API. Next to the checkbox there would be a link that says "show key". The link would only be available to administrator. This link would reveal a randomly generated 24 character key, something like ADRpWzFhMUd5e6NpRemMDEmJ. Maybe there should be a way to regenerate the key and/or set it manually, but not for an MVP. This is a mockup of what the UI could look like:

![github_updater](https://cloud.githubusercontent.com/assets/902911/12742691/39237e12-c999-11e5-8582-e4cc8a7d6e89.png)

Then, the actual api would be available at the following url, and to access it one would have to pass the key to it, like so:

```
<wordpress_base_url>/wp-plugins/github-updater/api.php?key=ADRpWzFhMUd5e6NpRemMDEmJ
```

In the file `api.php` we could not assume that any wordpress cookie was set, of course, and we would not be able to rely on any other "standard" wordpress authentication mechanism. We would however be able to check if the key was passed properly. This check would be the first thing done in that file, and if the correct key was not set we would `die` hard. This was we would ensure that security wasn't compromised by potential sloppy programming further down in the file (or dependencies of the file).

To actually perform an update of a theme or plugins, one would POST to an url like:

```
<wordpress_base_url>/wp-plugins/github-updater/api.php/update?key=ADRpWzFhMUd5e6NpRemMDEmJ&plugin=my/plugin
```

or

```
<wordpress_base_url>/wp-plugins/github-updater/api.php/update?key=ADRpWzFhMUd5e6NpRemMDEmJ&theme=my/theme
```

Maybe there should be a way to get information about current updates using GET:

```
<wordpress_base_url>/wp-plugins/github-updater/api.php/available_updates?key=ADRpWzFhMUd5e6NpRemMDEmJ
```

And maybe there should also be a way to update all available updates. I'm not sure exactly what the API should be, I'll let it evolve as I implement it and then we can have a discussion about it afterwards.

Sounds good?

**Comment by flashadvocate** [comment]
Definitely some merit and benefits to taking this route. It would be a good idea to ensure that the plugin doesn't return any fatal errors on update, so you don't lock yourself out of the admin panel with no way to get back in without deleting the errant plugin (or theme).

**Comment by limikael** [comment]
Sounds good!

Alright created a fork:

https://github.com/limikael/github-updater

Will start working, but the work will be a bit slow... :) Will try to release early, release often though, and submit a PR once I have something working!

**Comment by afragen** [comment]
I'll create a separate branch to submit the PR.

**Comment by afragen** [comment]
@limikael I've created a new branch `feature/rest-api`. Please submit all related PRs there. :+1: 

Also, I would think including the REST API components on a default basis and just showing/changing the key might be a better way to implement.

**Comment by afragen** [comment]
@flashadvocate I don't think it's really possible to check for a fatal on an active plugin upgrade, unless core does this already.

We should still be using core updating methods and we would need to ensure that the renaming code gets called too.

**Comment by limikael** [comment]
@afragen Ok will submit PRs there! I like the idea of having it enabled by default. This also means less code, so it makes more sense for an MVP. If someone would see a need to be able to turn it of, we could always add that functionality later.

**Comment by afragen** [comment]
No sure of the need to turn it off. If on and not used it doesn't do anything but create endpoints. :smile:

**Comment by afragen** [comment]
This was a split off of the feature plugin for Auth, https://github.com/georgestephanis/application-passwords/

**Comment by limikael** [comment]
Hi,

I have an initial version now, it is in this single commit:

https://github.com/limikael/github-updater/commit/9d68d2c48d32d20d461ea5362df53e150bf95486

I put it there in the spirit of "release early, release often", it's not something ready to merge yet, just something for getting feedback on. The way it works currently:

* The url that should be accessed is shown on the "Remote Management" tab.
* The parameters accepted by the url:
  * __key__ - The key as displayed on the "Remote Management". Must match the key stored on the
    system.
  * __plugin__ - Specify this to update a particular plugin. Should be specified as only the name of the
    plugin as it is known to the system, not the remote repo.
  * __theme__ - Specify this to update a theme.
  * __tag__ - Specify a particular tag, branch or commit for the update. If nothing is specified, it defaults
    to "master".
  * __committish__ - This is an alias for the tag parameter. I think that referring to a "committish" is
    more correct in in terms of git terminology, but it sounds a bit funny so this is why I decided to treat
    the committish and tag parameters as aliases for each other.
* Either the parameter __plugin__ or __theme__ must be specified. This means that this api only
  works for updating single plugins or themes, but I think this makes sense considering potential use
  cases.

Some points on the implementation:

* I did the authentication slightly different than previously discussed, it just shows the key on the "Remote Management" tab, as shown in the screenshot below.
* Currently it only works for GitHub plugins (not GitLab or BitBucket).
* It doesn't use much of the functionality in the Plugin and Theme classes. The needed functionality is "re-implemented" in the Base class.
* I haven't tested it extensively, but I have done some testing locally and it seems to work for both plugins and themes.
* It bypasses the "update needed" check, and it goes ahead and forces an update unconditionally. I
  think that this is a feature, so you can use it to update a themes and plugins you are working on,
  without having to update the version tag. I used code similar to the rollback functionality in the
  Theme and Plugin classes for this.

![screenshot from 2016-02-23 17-42-41](https://cloud.githubusercontent.com/assets/902911/13254977/f33b9d22-da54-11e5-9606-9c70a3a9ddc4.png)

Some questions:

* It prints out data using `json_encode` and the flag `JSON_PRETTY_PRINT`. This appeared in PHP
  5.4.0, is it ok to use it or would we want to be compatible with earlier versions?
* I don't think that having the functionality in the Base class is very elegant and not the way it should
  be. I have an idea for how to do it, but wanted to check with you guys first. For this discussion I will
  be describing theme updates, but the same changes would be done for plugins. What I'm thinking:
    * In the Theme class, there would be a function `update_single_theme($theme_name, $tag)` that
      would update a single theme.
    * In order to implement this function, I think it would be necessary to break out some of the code
      from the `get_theme_meta` function, since this functions loops over all themes. This would be
      done in a new function called `get_single_theme_meta`, and only the loop would remain in
      `get_theme_meta`, the rest of the functionality would go in `get_single_theme_meta`. The 
      `get_single_theme_meta` function would be used by both `get_theme_meta` and
      `update_single_theme`. Does this make sense and would it be a good solution?

All sorts of other feedback much welcomed!

**Comment by limikael** [comment]
Had some momentum, so I did in fact implement the solution I described before...

https://github.com/limikael/github-updater/commit/d92efc1b504845e40fd25c50c749fd38069ce702

**Comment by afragen** [comment]
@limikael rather than rewriting much of the existing classes it would be much better create separate classes for the REST API stuff and work on refactoring later. Otherwise, this will create too many issues with ongoing development.

**Comment by limikael** [comment]
Right...

So I'll create a separate class, and this class will manage the rest updates for both the plugins and the themes. Referred to in this discussion as the "RestUpdater". The RestUpdate will need to access information about the themes and plugins provided in the `$config` property in the Theme and Plugin classes. This property is public, so the RestUpdater can access it as `Plugin::instance()->config`. Maybe it's public "by accident", since it is not declared, but it's still public so this is a way to access it for now. Then there is the functionality that was before implemented in `get_remote_plugin_meta` which I moved to `get_single_remote_plugin_meta`. I'll copy paste this code and put in the RestUpdater?

**Comment by limikael** [comment]
Actually... The config property in Plugin and Theme is not accessible outside the class. I'll add the functions `get_plugin_configs` and `get_theme_configs` to those classes. Or should it be `get_plugin_config($plugin_name)`?

**Comment by limikael** [comment]
Done... I'm considering it to be an MVP now, so creating a pull request...

**Comment by afragen** [comment]
Merged.

**Comment by afragen** [comment]
@limikael what do you think of this, https://github.com/obenland/shiny-updates/issues/80

Any way to piggyback on that if it does get merged into core?

**Comment by afragen** [comment]
@limikael I've merged into `feature/rest-api` and made a few name changes to conform to the project style. Can you ensure that everything still works and I'll merge into `develop`?

Thanks.

**Comment by limikael** [comment]
@afragen I gave it a test and it seems to work fine. This is how I did the test, which can be considered a "smoke test" if someone else would like to test it in the future:

On an as clean as possible Wordpress installation:

1. I used github-updater to install a plugin from GitHub. In this case I used one of my plugins which is under development, namely this one: `https://github.com/limikael/wpcrud`.
2. I made a change to my plugin and pushed it to GitHub. I just added a comment somewhere, it doesn't need to have a new version specified, it should work anyway.
3. I went in to the github-updater settings page to check my REST update url on the "Remote Management" tab. In my case the url was `http://localhost/wordpress/wp-admin/admin-ajax.php?action=github-updater-update&key=76bb2b7c819c36ee37292b6978a4ad61`.
4. I added `&plugin=wpcrud` to the url and ran it from curl, the full command was: `curl "http://localhost/wordpress/wp-admin/admin-ajax.php?action=github-updater-update&key=76bb2b7c819c36ee37292b6978a4ad61&plugin=wpcrud"`. The output from the command was as follows, which looks ok.

        {
            "messages": [
                "Downloading update from <span class=\"code\">https:\/\/api.github.com\/repos\/limikael\/wpcrud\/zipball\/master<\/span>&#8230;",
                "Unpacking the update&#8230;",
                "Installing the latest version&#8230;",
                "Removing the old version of the plugin&#8230;",
                "Plugin updated successfully."
            ],
            "success": true
        }

5. I verified that the update I made in step 2. made it in to my Wordpress installation, which it did.

Looks ok?

**Comment by afragen** [comment]
Great! :100: 

Did the renaming work correctly?

I'll try to write up some instructions in the readme based upon what we've discussed here and we'll work on committing that from `develop`

**Comment by afragen** [comment]
Merged into `develop` 👍

Now we need some docs.

**Comment by limikael** [comment]
How do you mean did the renaming work? I did test the latest code that had the renamed classes in it, like `Rest_Update` and `Rest_Upgrader_Skin`, and it did seem to work for me. Is that what you mean or something else?

Yep some docs we need. I wrote a bit in this comment in this same thread: https://github.com/afragen/github-updater/issues/354#issuecomment-187744565. Could that be something to start with? I could take that and clean it up a bit and put it under "Remote Management Services" in the README.md file?

**Comment by afragen** [comment]
I meant did the updated plugin/theme have the correct directory renaming as opposed to what the download direct from GitHub looks like.

**Comment by limikael** [comment]
Aha... Tested a bit it what happens is the following, for the plugin `https://github.com/limikael/wpcrud`:

When I first install it using the UI of github-updater it will put itself in the directory `.../wp-content/plugins/wpcrud`. If I then do an update using the REST api and curl, the directory will be renamed to `.../wp-content/plugins/limikael-wpcrud-bc48f99`. The `bc48f99` corresponds to the latest commit. If I then push an update to git for the plugin and update it again using the rest API and curl, then the new version of the plugin will put itself in `.../wp-content/plugins/limikael-wpcrud-670e690`. And `670e690` is the new latest commit, so it seems good.

That's how it should behave, or?

**Comment by afragen** [comment]
No not good. It should be consistently named `../wp-content/plugins/wpcrud`

That means the endpoint isn't hitting the renaming code in the main plugin. `Base::upgrader_source_selection()` needs to run.

**Comment by limikael** [comment]
Darn... Ok let me look at it...

**Comment by afragen** [comment]
I think I fixed it. Update to latest `develop` branch and test please.
Just needed call `$this->load_post_filters()` which loads renaming code.

**Comment by limikael** [comment]
Ok verified, it kept its directory name on update!

If I write some docs, should I do that on my develop branch and PR it?

**Comment by limikael** [comment]
Wrote some docs and PRed to the rest-api branch....

https://github.com/afragen/github-updater/pull/400

**Comment by afragen** [comment]
> Ok verified, it kept its directory name on update!

Good, I'm glad it was a simple fix.

> Wrote some docs and PRed to the rest-api branch....
> 
> #400

PR to my `develop` branch is perfect.

**Comment by afragen** [comment]
I did a big refactor of `get_remote_{plugin|theme}_meta()`. It's now `get_remote_repo_meta()`

I also found out that renaming needed to change for Shiny Updates. This slightly affected the change I made for renaming in `Rest_Updater`, can you please test that REST API renaming works correctly again? Thanks.

**Comment by limikael** [comment]
Yep still works and the directory got renamed to its original name.

To be honest though, I only tested updating a plugin, I didn't test with a theme. Are you 99.999% sure that the fact that updating plugins work, also means that updating themes will work as well, or would you like me to test that too?

For the record, what is Shiny Updates?

**Comment by afragen** [comment]
I tested on both plugins/themes and I plan on testing more today. Yes, 99% sure it should work. :wink: 

Shiny Updates is the project for AJAX updates hoping to be introduced in WP 4.6. Because it's using AJAX and there are no page refreshes, it's much faster.

https:// github.com/obenland/shiny-updates

**Comment by limikael** [comment]
Cool!

**Comment by afragen** [comment]
I was able to test locally and hot damn!! 
Works great, plugins and themes. Renaming functioning.

Great work!! Thanks! 💯

**Comment by afragen** [comment]
Merged https://github.com/afragen/github-updater/commit/2773c2bb634257b9dc10d87d80ce65b2aa7dde13
